.clj-kondo/hooks/clojure/core.clj

```diff
@@ -179,6 +179,134 @@
     (non-thread-safe-form-should-end-with-exclamation* node))
   {:node node})
 
+;; ----------------- Modules -----------------
+
+;; TODO -- this hard coding is not ideal, it would be better to support metadata on the namespaces themselves.
+(def ^:private global-modules
+  "Global module namespaces can be imported anywhere, but their (non-module) internal namespaces are private.
+  We can nest global modules, in which case each module hides its internal even from its parent module, but is itself
+  importable even outside its parent module(s)."
+  ;; We keep this map sorted so parents precede their children. For legibility please keep them listed that way too..
+  (sorted-set
+   'metabase.db
+   ;; We may aspire to encapsulating more of these - either through refactoring or ^:clj-kondo/ignore.
+   'metabase.db.connection
+   'metabase.db.data-source
+   ;; This case may justify a "friends" style relationship with the metabase.driver module.
+   'metabase.db.jdbc-protocols
+   'metabase.db.metadata-queries
+   'metabase.db.query
+   'metabase.db.setup
+   ;; The namespaces only have their modularity broken within our tests, perhaps we should special case them.
+   'metabase.db.liquibase
+   'metabase.db.schema-migrations-test.impl
+   'metabase.db.test-util
+
+   'metabase.upload
+   ;; We anticipate using this CLJC ns for type inference on the frontend, so it's exposed outside the CLJ module.
+   'metabase.upload.types))
+
+;; TODO -- this hard coding is not ideal, it would be better to support metadata on the namespaces themselves.
+(def ^:private internal-modules
```

Comment:
We don't need these for uploads, but I think they're pretty natural, and only a little bit more complicated than global nested modules to support. A drop in the bucket compared to the vanilla machinery.